File: AndroidSerialization2/app/src/main/java/gmlim/sample/androidserialization2/MainActivity.java
package gmlim.sample.androidserialization2;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.util.ArrayList;

import android.app.Activity;
import android.os.Bundle;
import android.os.Parcel;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import android.widget.TextView;

public class MainActivity extends Activity {

    private LinearLayout results;
    private ScrollView scroll;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        results = (LinearLayout) findViewById(R.id.results);
        scroll = (ScrollView) findViewById(R.id.scroll);
    }

    public void testParcelable(View view) {
        TreeNode root = createNode(0);
        Parcel parcel = Parcel.obtain();

        long start = System.nanoTime();

        root.writeToParcel(parcel, 0);

        long finish = System.nanoTime();

        int length = parcel.marshall().length;
        parcel.setDataPosition(0); // reset for reading

        long start2 = System.nanoTime();

        TreeNode restored = TreeNode.CREATOR.createFromParcel(parcel);

        long finish2 = System.nanoTime();

        addResult("parcel: " + (finish - start) / 1_000_000 + "ms; unparcel: " + (finish2 - start2)
                / 1_000_000 + "ms; size: " + length);

        System.out.println(restored);

        parcel.recycle();

    }

    public void testSerializable(View view) {
        TreeNode root = createNode(0);

        ObjectOutputStream out = null;
        ByteArrayOutputStream bas = new ByteArrayOutputStream(1_000_000);

        long start = System.nanoTime();
        try {
            out = new ObjectOutputStream(bas);
            out.writeObject(root);

        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            if (out != null) {
                try {
                    out.close();
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        }

        long finish = System.nanoTime();

        byte[] byteArray = bas.toByteArray();
        int length = byteArray.length;
        ByteArrayInputStream bis = new ByteArrayInputStream(byteArray);
        ObjectInputStream in = null;
        TreeNode restored = null;

        long start2 = System.nanoTime();

        try {
            in = new ObjectInputStream(bis);

            restored = (TreeNode) in.readObject();

        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            if (in != null) {
                try {
                    in.close();
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        }

        long finish2 = System.nanoTime();

        addResult("serialize: " + (finish - start) / 1_000_000 + "ms; deserialize: "
                + (finish2 - start2) / 1_000_000 + "ms; size: " + length);

        System.out.println(restored);
    }

    private TreeNode createNode(int level) {
        if (level < 4) {
            return createRootNode(level + 1);
        } else {
            return createSimpleNode();
        }

    }

    private TreeNode createRootNode(int level) {
        TreeNode root = createSimpleNode();
        root.children = new ArrayList<TreeNode>(10);
        for (int i = 0; i < 10; i++) {
            root.children.add(createNode(level));
        }
        return root;
    }

    private TreeNode createSimpleNode() {
        TreeNode root = new TreeNode();
        root.string0 = "aaaaaaaaaa";
        root.string1 = "bbbbbbbbbb";
        root.string2 = "cccccccccc";
        root.int0 = 111111111;
        root.int1 = 222222222;
        root.int2 = 333333333;
        root.boolean0 = true;
        root.boolean1 = false;
        root.boolean2 = true;
        return root;
    }

    private void addResult(String message) {
        TextView result = new TextView(this);
        result.setText(message);
        result.setPadding(10, 10, 10, 10);
        results.addView(result);
        scroll.post(new Runnable() {
            @Override
            public void run() {
                scroll.scrollTo(0, scroll.getBottom());
            }
        });
    }
}

File: AndroidSerialization/app/src/main/java/gmlim/sample/androidserialization/MainActivity.kt
package gmlim.sample.androidserialization

import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.os.Parcel
import android.util.Log
import android.view.View
import android.widget.TextView
import kotlinx.android.synthetic.main.activity_main.*
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.ObjectInputStream
import java.io.ObjectOutputStream
import java.util.ArrayList

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
    }

    fun testParcelable(view: View) {
        val root = createNode(0)
        val parcel = Parcel.obtain()

        val start = System.nanoTime()

        root.writeToParcel(parcel, 0)

        val finish = System.nanoTime()

        val length = parcel.marshall().size
        parcel.setDataPosition(0) // reset for reading

        val start2 = System.nanoTime()

        val restored = TreeNode.CREATOR.createFromParcel(parcel)

        val finish2 = System.nanoTime()

        addResult("parcel: " + (finish - start) / 1000000 + "ms; unparcel: " + (finish2 - start2) / 1000000 + "ms; size: " + length)

        Log.d("stephen.lim", restored.toString())

        parcel.recycle()

    }

    fun testSerializable(view: View) {
        val root = createNode(0)

        var out: ObjectOutputStream? = null
        val bas = ByteArrayOutputStream(1000000)

        val start = System.nanoTime()
        try {
            out = ObjectOutputStream(bas)
            out!!.writeObject(root)

        } catch (e: Exception) {
            e.printStackTrace()
        } finally {
            if (out != null) {
                try {
                    out!!.close()
                } catch (e: Exception) {
                    e.printStackTrace()
                }

            }
        }

        val finish = System.nanoTime()

        val byteArray = bas.toByteArray()
        val length = byteArray.size
        val bis = ByteArrayInputStream(byteArray)
        var `in`: ObjectInputStream? = null
        var restored: TreeNode? = null

        val start2 = System.nanoTime()

        try {
            `in` = ObjectInputStream(bis)

            restored = `in`!!.readObject() as TreeNode

        } catch (e: Exception) {
            e.printStackTrace()
        } finally {
            if (`in` != null) {
                try {
                    `in`!!.close()
                } catch (e: Exception) {
                    e.printStackTrace()
                }

            }
        }

        val finish2 = System.nanoTime()

        addResult(
            "serialize: " + (finish - start) / 1000000 + "ms; deserialize: "
                    + (finish2 - start2) / 1000000 + "ms; size: " + length
        )

        Log.d("stephen.lim", restored.toString())
    }

    private fun createNode(level: Int): TreeNode {
        return if (level < 4) {
            createRootNode(level + 1)
        } else {
            createSimpleNode()
        }

    }

    private fun createRootNode(level: Int): TreeNode {
        val root = createSimpleNode()
        root.children = ArrayList<TreeNode>(10)
        for (i in 0..9) {
            (root.children as ArrayList<TreeNode>).add(createNode(level))
        }
        return root
    }

    private fun createSimpleNode(): TreeNode {
        val root = TreeNode()
        root.string0 = "aaaaaaaaaa"
        root.string1 = "bbbbbbbbbb"
        root.string2 = "cccccccccc"
        root.int0 = 111111111
        root.int1 = 222222222
        root.int2 = 333333333
        root.boolean0 = true
        root.boolean1 = false
        root.boolean2 = true
        return root
    }

    private fun addResult(message: String) {
        val result = TextView(this)
        result.text = message
        result.setPadding(10, 10, 10, 10)
        results.addView(result)
        scroll.post(Runnable { scroll.scrollTo(0, scroll.getBottom()) })
    }
}
